### PR TITLE
Proposal containing the results of the SilkX experiment

### PR DESCRIPTION
**Pull request by Perksey**
This shall remain in draft until the SilkX experiment has concluded!

### CODE REVIEW

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Khitiara) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:73]: backwards

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Khitiara) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:77]: clarify ability to return other Ptr-like types

Comment (Khitiara) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:77]: permit multiple-argument indexers on higher-arity Ptr types

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Khitiara) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:86]: still feel uneasy with implicit string conversions for allocation/perf reasons but that would need benchmarking and further analysis and consideration

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Khitiara) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:258]: I vote yes to this

**Review by Khitiara — COMMENTED** [review]
overall LGTM

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:258]: Yeah fair, just wasn't sure whether we wanted what is essentially a proxy to implement the interface itself, but yeah I don't see why not.

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:86]: If you don't want to marshal a string just don't pass a string 🤷

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:77]: I thought this was implied by `inner dimension`, but alright will reword

Comment (Khitiara) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:77]: yeah i just wanted that clarified in the wording, i think the parens around the `byte`/non-generic version hides it a bit at least to my reading

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:221]: Maxine has expressed interest in Discord in not doing this, and maintaining an INativeContext interface in some form rather than forcing delegates to be used.

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:73]:
```suggestion
`Ptr` and `Mut` `ref struct`s **shall** be defined wrapping a `ref readonly` and `ref` respectively. These types **shall** have generic and non-generic variants. This shall be repeated up to 3 pointer dimensions, taking into account C's mutability rules in each instantiation. This means that there will be 14 top-level generic structs and 14 top-level non-generic structs. These **should** be auto-generated to ensure they remain maintainable. IL weaving **may** be required for the higher dimensions.
```

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:77]:
```suggestion
- An indexer **must** be present accepting a `nuint` for the index, returning a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension (i.e. a `PtrPtr<T>`'s indexer will return a `Ptr<T>`) of type being indexed.
```

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:258]:
```suggestion
The binding class **shall** implement the static subinterface (i.e. to proxy calls to the *static default* for ease of use).
```

**Comment by uwx** [comment]
Notes from Discord chat:

> > > For this requirement, how do we implicitly handle this without a known length? For a C-string we can search for the null-terminator, but not all strings will necessarily follow that pattern, and not every byte ptr will be a c-string leading to a crash when VS inspects it. This also feels like something that should be in an extension method since really it should only be defined for byte, sbyte, short, ushort, int, uint as other generic types will not convert nicely. 
> > 
> > what is the motivation for the implicit operator? having to do new string(x) makes more sense to me
> > like, in C++ there is no implicit cast from char* to std::string
> > because the process of that conversion is not determinable
> > you have a pointer to memory, and i think it would be unwise to make assumptions about what that pointer means
> 
> ok lets make it an explicit operator for now 

Also, need a thorough discussion about left-to-right order of PtrPtrPtr chains, and what Mut and Ptr should be named for clarity

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (curin) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:76]: Regarding dimension mutability, more explicit definition maybe good. As of right now inner-most is first and outermost is last. This may also not be the most intuitive as I read it in the reverse. So MutPtr<byte> would be byte** const rather than byte * const *.

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (curin) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:86]: Maxine brought this up in the discord, this may be better as an explicit operator as assuming a Ptr<T> equates to a string is dangerous.

Comment (curin) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:86]: It is also probably good to limit this functionality to only String types (byte, sbyte, ushort, short, uint, int). Using extensions would work for a method, but we can't directly override an operator. It may be best to use an explicit function like ReadToString(), then the operator can call the function and we override that for specific types otherwise throw an exception.

Comment (Perksey) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:86]: Agreed, I think we could define `ReadToString` in `SilkMarshal` as an extension method for `Ptr`, `Mut`, `Ptr<T>`, and `Mut<T>` where T is one of: byte, sbyte, ushort, short, char, uint, int

File: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md
Comment (Perksey) [inline documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:75]: In the case of a generic pointer type, implicitly convertible to its non-generic counterpart?

Comment (curin) [reply documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md:75]: This should be easy enough as we would just be passing the InteriorRef as the Ref for the non-generic consturctor.

**Comment by Perksey** [comment]
# Maths

- We agree with the addition to add scalar operations over vectors (i.e. Vector4<T> * T does X*T, Y*T, ...)
- Add an analyser for encouraging the most correct and most efficient type instead of using sub-optimal types. 
- Ensure we've documented that.
- Vector * Matrix?
    - assuming row major, not allowing matrix * vector (only vector * matrix)
    - does it make sense to just do it the way game engines do it and the way 2.0 does it?
    - possibly technically invalid mathematically, even if common programmatically.
    - easy to hit corner case that not all users may understand.
- Ensuring we have no upcasting and downcasting operators that could have unexpected behaviour.
- System.Numerics is strictly right hand, which is the same as XNA and somewhat default for OpenGL users. This differs with DirectX that is typically left hand.
    - Happy with having explicit LH and RH functions.
- Missing matrix APIs that were added in .NET 8 to System.Numerics
    - CreateLookTo (additional to CreateLookAt)
    - CreateViewport
    - CreateBillboardLH
    - CreateConstrainedBillboardLH
    - CreateBillboardRH
    - CreateConstrainedBillboardRH
- Investigate newer less-hated alternatives to quaternions for future (see Freya Holmer talks)
- Possibly have a general Transform struct in the future?
- Rectangles and boxes are obscured with regards to internal layout, bindings to have interchange types that implicitly cast to the mathematical constructs that are represented by the Box/Rectangle types.
- Investigate in the future if we can either add an Origin property for compatibility or an analyser for users to know that Min is the new Origin.
- Approved notwithstanding the missing APIs.

**Comment by Perksey** [comment]
# SilkTouchX

- We discussed a particular problematic case where RegisterClassEx returns an atom which is later reinterpreted to be a pointer - a debugger will explode when inspecting this pointer as it is not necessarily 
- Where ReadOnlySpan<char> represents a string and we don't just want to pass the ref as-is (i.e. we want to add the null terminator like we do for string).
- Generally we think that providing a tool that works 90% of the time is fine, the unsafe overloads are always there, but we'd worry about users making incorrect assumptions and we can probably do implicit behaviour for that final 10%.
- Require that users manually encoding strings add that null terminator and document this. Our implicit ones do the right thing.
- Don't allow ref types to throw an error when being handed something that isn't a valid pointer.
- Approved provided that we:
    - make unsafe available
    - special case ROSpan<char> as above
- Future discussions need to be had on Vulkan implementation intricacies (getProcAddr) and also the addition of "complex" overloads.

**Comment by Perksey** [comment]
@curin @uwx @khitiara If any of you want to put up a PR for the CreateLookTo (additional to CreateLookAt), CreateViewport, CreateBillboardLH, CreateConstrainedBillboardLH, CreateBillboardRH, CreateConstrainedBillboardRH it'd be much appreciated sop we can get this in.

I'll update SilkTouchX now.

**Comment by Perksey** [comment]
@dotnet/silk-dotnet @uwx @Khitiara This is ready for merge I think.

**Review by Khitiara — APPROVED** [review]
LGTM

**Comment by Perksey** [comment]
Is it possible to get an approval here? The revisions to the proposals are as discussed in the WG meeting.

**Review by curin — APPROVED** [review]

**Comment by Perksey** [comment]
NOOOOOO NOT THE NATIVE BUILD JOBS

**Comment by Perksey** [comment]
Build is not triggering because there have been no source changes, merging manually.

File: documentation/proposals/Proposal - Generic Math.md
Comment (alexrp) [inline documentation/proposals/Proposal - Generic Math.md:383]: @Perksey don't suppose you have a link? Quaternions are, of course, ontologically evil, so I would be interested to hear about viable alternatives. :eyes:

Comment (Perksey) [reply documentation/proposals/Proposal - Generic Math.md:383]: I think it was @HurricanKai or @ThomasMi that raised this point (and alluded to things like rotors as well)

**Pull request merged**
